In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data=[['milk','bread','rice','book'],
      ['bread','jam','book','pen'],
      ['jam','milk','bread','rice','eggs'],
      ['rice','eggs','pen','book'],
     ['eggs','pen','milk','bread','jam'],
      ['eggs','rice','bread','jam']]

In [3]:
#        !pip install mlxtend


In [4]:
# Convert the data into a one-hot encoded DataFrame

from mlxtend.preprocessing import TransactionEncoder
te=TransactionEncoder()
te_array=te.fit_transform(data)

In [5]:
df=pd.DataFrame(te_array,columns=te.columns_)
df

book  bread   eggs    jam   milk    pen   rice
0   True   True  False  False   True  False   True
1   True   True  False   True  False   True  False
2  False   True   True   True   True  False   True
3   True  False   True  False  False   True   True
4  False   True   True   True   True   True  False
5  False   True   True   True  False  False   True

In [6]:
# Generate frequent itemsets with a minimum support threshold of 0.6 

from mlxtend.frequent_patterns import apriori
itemset=apriori(df,min_support=0.6,use_colnames=True)
itemset

support      itemsets
0  0.833333       (bread)
1  0.666667        (eggs)
2  0.666667         (jam)
3  0.666667        (rice)
4  0.666667  (jam, bread)

In [7]:
# Generate association rules with a minimum confidence threshold of 0.6

from mlxtend.frequent_patterns import association_rules
res=association_rules(itemset,metric='confidence',min_threshold=0.6)
res

antecedents consequents  antecedent support  consequent support   support  \
0       (jam)     (bread)            0.666667            0.833333  0.666667   
1     (bread)       (jam)            0.833333            0.666667  0.666667   

   confidence  lift  leverage  conviction  zhangs_metric  
0         1.0   1.2  0.111111         inf            0.5  
1         0.8   1.2  0.111111    1.666667            1.0

In [8]:
# Filter the results to show only relevant columns

results=res[['antecedents','consequents','support','confidence','lift']]
results

antecedents consequents   support  confidence  lift
0       (jam)     (bread)  0.666667         1.0   1.2
1     (bread)       (jam)  0.666667         0.8   1.2

Rule 1: {jam} -> {bread}
- Support: 0.666667 (66.67%)
  - This means that 66.67% of all transactions contain both jam and bread.
- Confidence: 1.0 (100%)
  - This means that 100% of transactions that contain jam also contain bread.
- Lift: 1.2
  - A lift value greater than 1 indicates a positive association between jam and bread. Specifically, the likelihood of buying bread increases by 20% when jam is purchased.

Rule 2: {bread} -> {jam}
- Support: 0.666667 (66.67%)
  - This means that 66.67% of all transactions contain both bread and jam.
- Confidence: 0.8 (80%)
  - This means that 80% of transactions that contain bread also contain jam.
- Lift: 1.2
  - A lift value greater than 1 indicates a positive association between bread and jam. Specifically, the likelihood of buying jam increases by 20% when bread is purchased.

**Key Insights**
      
1.Strong Association Between Jam and Bread:
  - Both rules show a strong positive association between jam and bread.
  - The high confidence values (100% and 80%) indicate that these items are frequently purchased together.
2.Directional Relationship:
  - The relationship is not symmetric:
    - If a customer buys jam, they are 100% likely to also buy bread.
    - If a customer buys bread, they are 80% likely to also buy jam.
3.Lift Interpretation:
  - The lift value of 1.2 for both rules indicates that the presence of one item (jam or bread) increases the likelihood of purchasing the other item by 20%.